EIPS/eip-747.md

```diff
@@ -64,51 +52,44 @@ If the asset is already listed in the user's wallet, the wallet SHOULD re-fetch
 - Downloading the image failed to load
 - The wallet blocked the asset, either due to an allowlist or denylist
 
-## Rationale
+### `ERC1046` type
 
-Displaying a user's assets is a basic feature that every modern dapp user expects. However, keeping this list, and polling for it from the network can be costly, especially on bandwidth constrained devices.
+The format of the options field is:
 
-Most wallets today either manage their own assets list, which they store client side, or they query a centralized API for balances, which reduces decentralization, letting that API's owner easily correlate account holders with their IP addresses.
+```typescript
+interface ERC1046WatchAssetOptions {
+{
+    address: string; // The hexadecimal Ethereum address of the token contract
+    chainId?: number; // The chain ID of the asset. If empty, defaults to the current chain ID.
+  };
+}
+```
 
-Maintaining one of these assets lists becomes a political act, and maintainers can be subject to regular harassment and pressure to list otherwise unknown assets.
+`address` is required, and the other fields are optional. `address` MUST be the hexadecimal Ethereum address of the token contract. `chainId` MUST be the chain ID to which the asset belongs.
 
-Furthermore, automatically listing assets makes assets into a sort of spam mail: Users suddenly seeing new assets that they don't care about in their wallet can be used to bombard them with information that they didn't opt into.
+`wallet_watchAsset` MUST fetch the [ERC-1046](./eip-1046.md) `tokenURI` and check the `interop` field to determine the type of the token. If the parsing fails, or the type is unknown, the RPC call MUST error.
 
-This phenomenon is exacerbated by the trend towards airdropped tokens, which has been a cause of network congestion, because spamming people with new tokens has so far been rewarded with user attention.
+If the wallet does not recognize the `chainId`, or the `chainId` is blank and the wallet does not have a concept of "active" chain, the call MUST fail.
```

Comment:
That makes sense. I'll make that change.

Reply:
Actually, I disagree here. I feel that being able to tell whether the user was prompted (and if not, why not) is useful.

Reply:
What is the use case for the dapp, and is it worth doxxing the user a little bit?  For example, you might be able to use this to figure out whether the user is using a specific wallet, or is following a specific blacklist, or to identify whether your scam was successful in prompting the user or not.

While it is possible there is a strong use case for giving a response to the dapp, my instinct is that the use cases don't outweigh the costs here.

Reply:
I feel like if there's an actionable error, that it should be communicated to the DApp. For example, if the chain ID is missing, the DApp can request to add it.

Reply:
I believe most (all?) such errors are either programming errors or there are better ways to prevent them in the dapp.  Programming errors in the dapp should be caught and dealt with long before the dapp makes it into production, and errors like the one you suggested should be dealt with by first verifying the chain you want has been added.  I don't think we should doxx every single user of every wallet because some dapps may have programming errors or follows a bad flow.

In the example you provided, the dapp should first add the chain and *then* add the token.  For example:
```ts
await eth_addChain(5)
await eth_watchAsset('ERC1046', { address, chainId: 5 })
```

If the user declines the chain addition, I don't think the wallet should tell the user (it isn't any of the dapp's business what chains the user is using) and the watchAsset will then proceed the same either way from the dapp's point of view.

In your example, the dapp could use this to probe what chains the wallet has configured which may be information the user prefers not to be shared with the dapp.

Reply:
Okay, you've convinced me. I'll add that.

Reply:
Actually, wait, never mind. This definitely should be an error because of automated error reporting tools. If there's an error in the flow, then it should be identified instead of silently failing.

Reply:
> Actually, wait, never mind. This definitely should be an error because of automated error reporting tools. If there's an error in the flow, then it should be identified instead of silently failing.

It is an "error in the flow" only if you consider `eth_watchAsset` to be an assertion rather than a suggestion.  I am of the opinion that a dapp saying `eth_watchAsset` is merely a suggestion.  It is the dapp saying something like:
> Hey wallet, we have some information that suggests the user might care about this asset, consider adding it to their portfolio.

I am pretty strongly against this function being a demand like:
> Hey wallet, add this asset to the user's portfolio, if you don't the app won't work and we tell the user to find a new wallet.

Reply:
> I am pretty strongly against this function being a demand like 'Hey wallet, add this asset to the user's portfolio, if you don't the app won't work and we tell the user to find a new wallet.'

Rationale, please? I don't see any reason why this shouldn't be an assertion.

Reply:
I don't see any reason why it *should* be an assertion.  What is a situation where the dapp must know what assets the user is tracking?  What is the benefit to the *user* for divulging that information to the dapp?  Is that benefit worth it to doxx and fingerprint the user to every dapp out there?

For privacy reasons, we should default to giving less information to dapps, not more, unless there is a compelling reason to doxx the user, so the onus (IMO) is on the person proposing to standardize the revealing of additional information to the dapp to argue why it is necessary

Reply:
> What is a situation where the dapp must know what assets the user is tracking?

I can't think of one. Fair enough.